src/EditorFeatures/CSharpTest/CodeActions/ConvertLinq/ConvertLinqQueryToForEachTests.cs

```diff
@@ -0,0 +1,2149 @@
+using System.Linq;
+class C
+{
+    void M(IEnumerable<int> nums)
+    {
+        foreach (int n1 in nums)
+        {
+            foreach (int n2 in nums)
+            {
+            Console.WriteLine(n1);
+        }
+        }
+    }
+}
+";
+            await TestInRegularAndScriptAsync(source, output);
+        }
+
+        [Fact]
+        public async Task CallingMethodWithIEnumerable()
+        {
+            string source = @"
+using System;
+using System.Collections.Generic;
+using System.Linq;
+class C
+{
+    void M(IEnumerable<int> nums)
+    {
+        var q = [||]from int n1 in nums 
+                from int n2 in nums
+                select n1;
+        N(q);
+    }
+
+    void N(IEnumerable<int> q) {}
+}
+";
+
+            string output = @"
+using System;
+using System.Collections.Generic;
+using System.Linq;
+class C
+{
+    void M(IEnumerable<int> nums)
+    {
+        IEnumerable<int> localFunction()
+        {
+            foreach (int n1 in nums)
+            {
+                foreach (int n2 in nums)
+                {
+                    yield return n1;
+                }
+            }
+        }
+
+        var q = localFunction();
+        N(q);
+    }
+
+    void N(IEnumerable<int> q) {}
+}
+";
+            await TestInRegularAndScriptAsync(source, output);
+        }
+
+
+        [Fact]
+        public async Task AssignmentExpression()
+        {
+            string source = @"
+using System;
+using System.Collections.Generic;
+using System.Linq;
+class C
+{
+    void M(IEnumerable<int> nums)
+    {
+        IEnumerable<int> q;
+        q = [||]from int n1 in nums 
+                from int n2 in nums
+                select n1;
+
+        N(q);
+    }
+
+    void N(IEnumerable<int> q) {}
+}
+";
+
+            string output = @"
+using System;
+using System.Collections.Generic;
+using System.Linq;
+class C
+{
+    void M(IEnumerable<int> nums)
+    {
+        IEnumerable<int> q;
+        IEnumerable<int> localFunction()
+        {
+            foreach (int n1 in nums)
+            {
+                foreach (int n2 in nums)
+                {
+                    yield return n1;
+                }
+            }
+        }
+
+        q = localFunction();
+
+        N(q);
+    }
+
+    void N(IEnumerable<int> q) {}
+}
+";
+            await TestInRegularAndScriptAsync(source, output);
+        }
+
+        [Fact]
+        public async Task Select()
+        {
+            string source = @"
+using System;
+using System.Collections.Generic;
+using System.Linq;
+class Query
+{
+    public static void Main(string[] args)
+    {
+        List<int> c = new List<int>{ 1, 2, 3, 4, 5, 6, 7 };
+        var r = [||]from i in c select i+1;
+    }
+}";
+            string output = @"
+using System;
+using System.Collections.Generic;
+using System.Linq;
+class Query
+{
+    public static void Main(string[] args)
+    {
+        List<int> c = new List<int>{ 1, 2, 3, 4, 5, 6, 7 };
+        IEnumerable<int> localFunction()
```

Comment:
this local function should probably be named contextually to some better.

Reply:
Any ideas?

---
In reply to: [175267354](https://github.com/dotnet/roslyn/pull/25362#discussion_r175267354) [](ancestors = 175267354)

Reply:
it looks like we have this.
http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/LanguageServices/SemanticsFactsService/ISemanticFactsService.cs,75

improve that method?

Reply:
Not sure I understand. Could you please explain?

---
In reply to: [176885977](https://github.com/dotnet/roslyn/pull/25362#discussion_r176885977) [](ancestors = 176885977)

Reply:
this method "GenerateNameForExpression" tries to create a contextual name from given syntax node. by looking at the code, it tries to see whether things are like variable, method call and etc, and tries to create name from it. you can add query expression support there so that it gives you good name like, if query expression ends with select n1. it can give back GetN1s or something like that rather than localFunction.

also, I believe we have humenizer or something that fix up plurals and stuff? not sure where that code is though.

Reply:
works great! thank you!

---
In reply to: [176894908](https://github.com/dotnet/roslyn/pull/25362#discussion_r176894908) [](ancestors = 176894908)